File: app/src/main/java/uqac/ca/isstracker/handlers/AstronautAdapter.java
package uqac.ca.isstracker.handlers;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.support.annotation.*;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.*;

import com.squareup.picasso.Picasso;

import java.util.*;

import uqac.ca.isstracker.model.Astronaut;
import uqac.ca.isstracker.R;

public class AstronautAdapter extends ArrayAdapter<Astronaut> {

    private Context mContext;
    private List<Astronaut> astronautsList = new ArrayList<>();

    private ImageView mProfileImage;
    private ImageView mCountryFlag;
    private ImageView mAstronautTwitter;
    private ImageView mAstronautWiki;
    private TextView mName;
    private TextView mRole;
    private TextView mDate;
    private TextView mBio;

    private int currentPosition = -1;

    public AstronautAdapter(@NonNull Context context, ArrayList<Astronaut> list) {
        super(context, 0 , list);
        mContext = context;
        astronautsList = list;
    }

    @Override
    public View getView(int position, @Nullable View convertView, @NonNull ViewGroup parent) {
        View listItem = convertView;
        currentPosition = position;
        if(listItem == null)
            listItem = LayoutInflater.from(mContext).inflate(R.layout.astronaut_row, parent, false);

        Astronaut currentAstronaut = astronautsList.get(currentPosition);

          mProfileImage = (ImageView) listItem.findViewById(R.id.img);
          mCountryFlag = (ImageView) listItem.findViewById(R.id.countryFlag);
          mAstronautTwitter = (ImageView) listItem.findViewById(R.id.astronautTwitter);
          mAstronautWiki = (ImageView) listItem.findViewById(R.id.astronautWiki);
          mName = (TextView) listItem.findViewById(R.id.name);
          mRole = (TextView) listItem.findViewById(R.id.role);
          mDate = (TextView) listItem.findViewById(R.id.days_since_launch);
          mBio = (TextView) listItem.findViewById(R.id.bio);


        Picasso.get().load(currentAstronaut.getImage()).into(mProfileImage);
        Picasso.get().load(currentAstronaut.getCountryLink()).into(mCountryFlag);

        mAstronautTwitter.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Astronaut currentAstronaut = astronautsList.get(currentPosition);
                if (currentAstronaut.getTwitter().length() != 0) {
                    Toast.makeText(mContext, currentAstronaut.getTwitter(), Toast.LENGTH_SHORT).show();
                }
            }
        });

        mAstronautWiki.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Astronaut currentAstronaut = astronautsList.get(currentPosition);
                if (currentAstronaut.getTwitter().length() != 0) {
                    Toast.makeText(mContext, currentAstronaut.getWiki(), Toast.LENGTH_SHORT).show();
                }
            }
        });

        mName.setText(currentAstronaut.getName());
        mRole.setText(currentAstronaut.getRole());
        mDate.setText(currentAstronaut.getLaunchDate());
        mBio.setText(currentAstronaut.getBio());

        return listItem;
    }
}

File: app/src/main/java/uqac/ca/isstracker/adapters/AstronautAdapter.kt
package uqac.ca.isstracker.adapters

import android.content.Context
import android.content.Intent
import android.net.Uri
import android.support.v7.widget.RecyclerView
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.*

import com.squareup.picasso.Picasso

import java.util.ArrayList

import uqac.ca.isstracker.R
import uqac.ca.isstracker.model.Astronaut

// Provide a suitable constructor (depends on the kind of dataset)
class AstronautAdapter(private val appContext: Context, private val dataset: ArrayList<Astronaut>) : RecyclerView.Adapter<AstronautAdapter.CardViewHolder>()
{
    // Provide a reference to the views for each data item
    class CardViewHolder(v: View) : RecyclerView.ViewHolder(v)
    {
        var name: TextView
        var role: TextView
        var bio: TextView
        var launchDate: TextView
        var profileImage: ImageView
        var countryFlag: ImageView
        var astronautTwitter: ImageView
        var astronautWiki: ImageView

        init
        {
            this.name = v.findViewById(R.id.name)
            this.role = v.findViewById(R.id.role)
            this.bio = v.findViewById(R.id.bio)
            this.launchDate = v.findViewById(R.id.launchDate)

            this.profileImage = v.findViewById(R.id.img)
            this.countryFlag = v.findViewById(R.id.countryFlag)
            this.astronautTwitter = v.findViewById(R.id.twitter)
            this.astronautWiki = v.findViewById(R.id.wiki)
        }
    }

    // Create new views (invoked by the layout manager)
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): CardViewHolder
    {
        // create a new view
        val v = LayoutInflater.from(parent.context)
                .inflate(R.layout.astro_card, parent, false)

        return CardViewHolder(v)
    }

    // Replace the contents of a view (invoked by the layout manager)
    override fun onBindViewHolder(holder: CardViewHolder, position: Int)
    {
        // - get element from dataset at this position
        val astro = dataset[position]

        // - replace the contents of the view with that element
        //Texts
        holder.name.text = astro.name
        holder.role.text = "Role: " + astro.role
        holder.launchDate.text = "Launch date: " + astro.launchDate
        holder.bio.text = astro.bio

        //Images
        Picasso.get().load(astro.image).into(holder.profileImage)
        Picasso.get().load(astro.countryLink).into(holder.countryFlag)

        //Social networks
        if (astro.twitter.length == 0)
        {
            holder.astronautTwitter.visibility = View.INVISIBLE
        }
        else
        {
            holder.astronautTwitter.setOnClickListener {
                Log.d("ACTION", astro.twitter)
                appContext.startActivity(Intent(Intent.ACTION_VIEW, Uri.parse(astro.twitter)))
            }
        }

        if (astro.wiki.length == 0)
        {
            holder.astronautWiki.visibility = View.INVISIBLE
        }
        else
        {
            holder.astronautWiki.setOnClickListener {
                Log.d("ACTION", astro.wiki)
                appContext.startActivity(Intent(Intent.ACTION_VIEW, Uri.parse(astro.wiki)))
            }
        }
    }

    // Return the size of dataset (invoked by the layout manager)
    override fun getItemCount(): Int
    {
        return dataset.size
    }
}
